// gatsby-source-datocms/hooks/sourceNodes/destroyEntityNode.js
"use strict";

var _require = require('humps'),
    pascalize = _require.pascalize;

var ENTITY_TO_NODE_IDS = {
  item_type: function item_type(entity) {
    return "DatoCmsModel-".concat(entity.id);
  },
  field: function field(entity) {
    return "DatoCmsField-".concat(entity.id);
  },
  upload: function upload(entity) {
    return "DatoCmsAsset-".concat(entity.id);
  },
  item: function item(entity, _ref) {
    var entitiesRepo = _ref.entitiesRepo;
    var siteEntity = entitiesRepo.findEntitiesOfType('site')[0];
    var type = pascalize(entity.itemType.apiKey);
    return siteEntity.locales.map(function (locale) {
      return "DatoCms".concat(type, "-").concat(entity.id, "-").concat(locale);
    });
  },
  site: function site(entity, _ref2) {
    var entitiesRepo = _ref2.entitiesRepo;
    return entity.locales.map(function (locale) {
      return "DatoCmsSite-".concat(entity.id, "-").concat(locale);
    });
  }
};

module.exports = function (entity, context) {
  if (!ENTITY_TO_NODE_IDS[entity.type]) {
    console.log("Don't know how to delete entity of type '".concat(entity.type, "'!"));
    return;
  }

  var result = ENTITY_TO_NODE_IDS[entity.type](entity, context);
  var nodeIdsToDelete = Array.isArray(result) ? result : [result];
  nodeIdsToDelete.map(function (nodeId) {
    var node = context.getNode(nodeId);

    if (node) {
      context.actions.deleteNode({
        node: node
      });
    }
  });
};
