// gatsby-source-datocms/hooks/sourceNodes/createTypes/fieldTypes/DatoCmsColorField.js
"use strict";

module.exports = function (_ref) {
  var actions = _ref.actions,
      schema = _ref.schema;
  actions.createTypes([schema.buildObjectType({
    name: 'DatoCmsColorField',
    extensions: {
      infer: false
    },
    fields: {
      red: 'Int',
      green: 'Int',
      blue: 'Int',
      alpha: 'Int',
      rgb: {
        type: 'String',
        resolve: function resolve(parent) {
          if (parent.alpha === 255) {
            return "rgb(".concat(parent.red, ", ").concat(parent.green, ", ").concat(parent.blue, ")");
          }

          return "rgba(".concat(parent.red, ", ").concat(parent.green, ", ").concat(parent.blue, ", ").concat(parent.alpha, ")");
        }
      },
      hex: {
        type: 'String',
        resolve: function resolve(parent) {
          var rgba = ['red', 'green', 'blue', 'alpha'].map(function (component) {
            var hex = parent[component].toString(16);
            return hex.length === 1 ? "0".concat(hex) : hex;
          });
          return rgba[3] === 'ff' ? "#".concat(rgba.slice(0, 3).join('')) : "#".concat(rgba.join(''));
        }
      }
    }
  })]);
};
