// jsonref/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("./errors");
const meta = require("./meta");
const ref_1 = require("./ref");
__export(require("./errors"));
var meta_1 = require("./meta");
exports.getMeta = meta_1.getMeta;
exports.isAnnotated = meta_1.isAnnotated;
exports.isRef = meta_1.isRef;
exports.normalize = meta_1.normalize;
exports.normalizeUri = meta_1.normalizeUri;
var pointer_1 = require("./pointer");
exports.pointer = pointer_1.resolve;
__export(require("./rebase"));
function scope(data) {
    if (meta.isAnnotated(data)) {
        return meta.getMeta(data).scope;
    }
}
exports.scope = scope;
async function parse(dataOrUri, opts) {
    let obj;
    if (!opts || !opts.scope) {
        throw new Error('No scope');
    }
    if (typeof dataOrUri === 'string') {
        if (!opts.retriever) {
            throw new Error('No retriever');
        }
        const uri = new URL(dataOrUri).toString();
        obj = await opts.retriever(uri);
        if (!opts.registry) {
            opts.registry = {};
        }
        if (uri !== opts.scope) {
            opts.registry[meta.normalizeUri(uri)] = obj;
        }
    }
    else if (dataOrUri === null || typeof dataOrUri !== 'object') {
        throw new TypeError('Bad data');
    }
    else {
        obj = dataOrUri;
    }
    if (meta.isAnnotated(obj)) {
        return obj;
    }
    else {
        meta.annotate(obj, opts);
        if (meta.getMeta(obj).refs.size > 0) {
            const missingRefs = meta.missingRefs(obj);
            if (missingRefs.length) {
                if (!opts.retriever) {
                    throw new Error('No retriever');
                }
                const registry = meta.getMeta(obj).registry;
                const errors = [];
                for (let r of missingRefs) {
                    try {
                        registry[r] = await opts.retriever(r);
                    }
                    catch (e) {
                        errors.push(new errors_1.RetrieverError(r, e));
                    }
                }
                if (errors.length) {
                    throw new errors_1.ParserError(meta.getMeta(obj).scope, 'retriever', errors);
                }
            }
            return ref_1.resolve(obj, opts);
        }
        else {
            return obj;
        }
    }
}
exports.parse = parse;
