// datocms-client/lib/contentfulImport/datoFieldValidatorsFor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _datoLinkItemTypeFor = _interopRequireDefault(require("./datoLinkItemTypeFor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var datoValidatorsForString = function datoValidatorsForString(_ref) {
  var field = _ref.field;
  var validators = {};

  if (field.required) {
    validators.required = {};
  }

  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = field.validations[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var validation = _step.value;

      if (Object.prototype.hasOwnProperty.call(validation, 'size')) {
        validators.length = {};

        if (validation.size.min && validation.size.min === validation.size.max) {
          validators.length.eq = validation.size.min.toString();
        } else {
          if (validation.size.min) {
            validators.length.min = validation.size.min.toString();
          }

          if (validation.size.max) {
            validators.length.max = validation.size.max.toString();
          }
        }
      }

      if (Object.prototype.hasOwnProperty.call(validation, 'unique')) {
        validators.unique = {};
      }

      if (Object.prototype.hasOwnProperty.call(validation, 'in')) {
        validators["enum"] = {
          values: validation["in"]
        };
      }

      if (Object.prototype.hasOwnProperty.call(validation, 'regexp')) {
        validators.format = {
          customPattern: validation.regexp.pattern
        };
      }
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"] != null) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  return validators;
};

var datoValidatorsForInteger = function datoValidatorsForInteger(_ref2) {
  var field = _ref2.field;
  var validators = {};

  if (field.required) {
    validators.required = {};
  }

  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    for (var _iterator2 = field.validations[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var validation = _step2.value;

      if (Object.prototype.hasOwnProperty.call(validation, 'range')) {
        validators.numberRange = {};

        if (validation.range.min) {
          validators.numberRange.min = validation.range.min.toString();
        }

        if (validation.range.max) {
          validators.numberRange.max = validation.range.max.toString();
        }
      }
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
        _iterator2["return"]();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  return validators;
};

var datoValidatorsForDate = function datoValidatorsForDate(_ref3) {
  var field = _ref3.field;
  var validators = {};

  if (field.required) {
    validators.required = {};
  }

  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    for (var _iterator3 = field.validations[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      var validation = _step3.value;

      if (Object.prototype.hasOwnProperty.call(validation, 'dateRange')) {
        validators.dateRange = {
          min: validation.dateRange.min,
          max: validation.dateRange.max
        };
      }
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3["return"] != null) {
        _iterator3["return"]();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }

  return validators;
};

var datoValidatorsForLocation = function datoValidatorsForLocation(_ref4) {
  var field = _ref4.field;
  var validators = {};

  if (field.required) {
    validators.required = {};
  }

  return validators;
};

var datoValidatorsForObject = function datoValidatorsForObject(_ref5) {
  var field = _ref5.field;
  var validators = {};

  if (field.required) {
    validators.required = {};
  }

  return validators;
};

var datoValidatorsForLink = function datoValidatorsForLink(_ref6) {
  var field, itemTypes, validators;
  return regeneratorRuntime.async(function datoValidatorsForLink$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          field = _ref6.field, itemTypes = _ref6.itemTypes;
          validators = {
            itemItemType: {
              itemTypes: (0, _datoLinkItemTypeFor["default"])({
                field: field,
                itemTypes: itemTypes
              })
            }
          };

          if (field.required) {
            validators.required = {};
          }

          return _context.abrupt("return", validators);

        case 4:
        case "end":
          return _context.stop();
      }
    }
  });
};

var datoValidatorsForAsset = function datoValidatorsForAsset(_ref7) {
  var field = _ref7.field;
  var validators = {};

  if (field.required) {
    validators.required = {};
  }

  var _iteratorNormalCompletion4 = true;
  var _didIteratorError4 = false;
  var _iteratorError4 = undefined;

  try {
    for (var _iterator4 = field.validations[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
      var validation = _step4.value;

      if (Object.prototype.hasOwnProperty.call(validation, 'assetFileSize')) {
        if (Object.prototype.hasOwnProperty.call(validation.assetFileSize, 'min')) {
          validators.file_size = _objectSpread({}, validators.file_size, {
            min_value: validation.assetFileSize.min,
            min_unit: 'B'
          });
        }

        if (Object.prototype.hasOwnProperty.call(validation.assetFileSize, 'max')) {
          validators.file_size = _objectSpread({}, validators.file_size, {
            max_value: validation.assetFileSize.max,
            max_unit: 'B'
          });
        }
      }
    }
  } catch (err) {
    _didIteratorError4 = true;
    _iteratorError4 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion4 && _iterator4["return"] != null) {
        _iterator4["return"]();
      }
    } finally {
      if (_didIteratorError4) {
        throw _iteratorError4;
      }
    }
  }

  return validators;
};

var datoValidatorsForArray = function datoValidatorsForArray(_ref8) {
  var field, itemTypes, validators, _iteratorNormalCompletion5, _didIteratorError5, _iteratorError5, _iterator5, _step5, validation;

  return regeneratorRuntime.async(function datoValidatorsForArray$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          field = _ref8.field, itemTypes = _ref8.itemTypes;
          validators = {};

          if (field.items.type === 'Link' && field.items.linkType === 'Entry') {
            validators = {
              itemsItemType: {
                itemTypes: (0, _datoLinkItemTypeFor["default"])({
                  field: field.items,
                  itemTypes: itemTypes
                })
              }
            };
          }

          _iteratorNormalCompletion5 = true;
          _didIteratorError5 = false;
          _iteratorError5 = undefined;
          _context2.prev = 6;

          for (_iterator5 = field.validations[Symbol.iterator](); !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {
            validation = _step5.value;

            if (Object.prototype.hasOwnProperty.call(validation, 'size')) {
              validators.size = {};

              if (validation.size.min && validation.size.min === validation.size.max) {
                validators.size.eq = validation.size.min.toString();
              } else {
                if (validation.size.min) {
                  validators.size.min = validation.size.min.toString();
                }

                if (validation.size.max) {
                  validators.size.max = validation.size.max.toString();
                }
              }
            }
          }

          _context2.next = 14;
          break;

        case 10:
          _context2.prev = 10;
          _context2.t0 = _context2["catch"](6);
          _didIteratorError5 = true;
          _iteratorError5 = _context2.t0;

        case 14:
          _context2.prev = 14;
          _context2.prev = 15;

          if (!_iteratorNormalCompletion5 && _iterator5["return"] != null) {
            _iterator5["return"]();
          }

        case 17:
          _context2.prev = 17;

          if (!_didIteratorError5) {
            _context2.next = 20;
            break;
          }

          throw _iteratorError5;

        case 20:
          return _context2.finish(17);

        case 21:
          return _context2.finish(14);

        case 22:
          return _context2.abrupt("return", validators);

        case 23:
        case "end":
          return _context2.stop();
      }
    }
  }, null, null, [[6, 10, 14, 22], [15,, 17, 21]]);
};

var _callee = function _callee(_ref9) {
  var field, itemTypes;
  return regeneratorRuntime.async(function _callee$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          field = _ref9.field, itemTypes = _ref9.itemTypes;
          _context3.t0 = field.type;
          _context3.next = _context3.t0 === 'Symbol' ? 4 : _context3.t0 === 'Text' ? 4 : _context3.t0 === 'Integer' ? 5 : _context3.t0 === 'Number' ? 5 : _context3.t0 === 'Date' ? 6 : _context3.t0 === 'Location' ? 7 : _context3.t0 === 'Object' ? 8 : _context3.t0 === 'Link' ? 9 : _context3.t0 === 'Array' ? 15 : _context3.t0 === 'Boolean' ? 16 : 16;
          break;

        case 4:
          return _context3.abrupt("return", datoValidatorsForString({
            field: field
          }));

        case 5:
          return _context3.abrupt("return", datoValidatorsForInteger({
            field: field
          }));

        case 6:
          return _context3.abrupt("return", datoValidatorsForDate({
            field: field
          }));

        case 7:
          return _context3.abrupt("return", datoValidatorsForLocation({
            field: field
          }));

        case 8:
          return _context3.abrupt("return", datoValidatorsForObject({
            field: field
          }));

        case 9:
          _context3.t1 = field.linkType;
          _context3.next = _context3.t1 === 'Entry' ? 12 : _context3.t1 === 'Asset' ? 13 : 14;
          break;

        case 12:
          return _context3.abrupt("return", datoValidatorsForLink({
            field: field,
            itemTypes: itemTypes
          }));

        case 13:
          return _context3.abrupt("return", datoValidatorsForAsset({
            field: field
          }));

        case 14:
          return _context3.abrupt("return", {});

        case 15:
          return _context3.abrupt("return", datoValidatorsForArray({
            field: field,
            itemTypes: itemTypes
          }));

        case 16:
          return _context3.abrupt("return", {});

        case 17:
        case "end":
          return _context3.stop();
      }
    }
  });
};

exports["default"] = _callee;
