// gatsby-source-datocms/hooks/sourceNodes/createTypes/item/fields/richText.js
"use strict";

var _require = require('humps'),
    camelize = _require.camelize,
    pascalize = _require.pascalize;

module.exports = function (_ref) {
  var parentItemType = _ref.parentItemType,
      field = _ref.field,
      schema = _ref.schema,
      gqlItemTypeName = _ref.gqlItemTypeName,
      entitiesRepo = _ref.entitiesRepo;
  var fieldKey = camelize(field.apiKey);
  var parentItemTypeName = gqlItemTypeName(parentItemType);
  var itemTypeIds = field.validators[field.fieldType === 'rich_text' ? 'richTextBlocks' : 'itemsItemType'].itemTypes;

  if (itemTypeIds.length === 0) {
    return {
      fieldType: 'String'
    };
  }

  if (itemTypeIds.length === 1) {
    var linkedItemType = entitiesRepo.findEntity('item_type', itemTypeIds[0]);
    return {
      fieldType: {
        type: "[".concat(gqlItemTypeName(linkedItemType), "]"),
        allLocalesResolver: function allLocalesResolver(parent) {
          return parent.value___NODE;
        },
        normalResolver: function normalResolver(parent) {
          return parent["".concat(fieldKey, "___NODE")];
        },
        resolveFromValue: function resolveFromValue(ids, args, context) {
          return context.nodeModel.getNodesByIds({
            ids: ids
          });
        }
      }
    };
  }

  var unionType = "DatoCmsUnionFor".concat(parentItemTypeName).concat(pascalize(field.apiKey));
  return {
    types: [schema.buildUnionType({
      name: unionType,
      types: itemTypeIds.map(function (id) {
        return gqlItemTypeName(entitiesRepo.findEntity('item_type', id));
      })
    })],
    fieldType: {
      type: "[".concat(unionType, "]"),
      allLocalesResolver: function allLocalesResolver(parent) {
        return parent.value___NODE;
      },
      normalResolver: function normalResolver(parent) {
        return parent["".concat(fieldKey, "___NODE")];
      },
      resolveFromValue: function resolveFromValue(ids, args, context) {
        return context.nodeModel.getNodesByIds({
          ids: ids
        });
      }
    }
  };
};
