// datocms-client/lib/wpImport/import/authors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = authors;

var _allPages = _interopRequireDefault(require("../utils/allPages"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _require = require('../utils/progress'),
    progress = _require.progress;

function authors(dato, wp, itemTypeId) {
  var mapping, resources, tick, _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, author;

  return regeneratorRuntime.async(function authors$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          mapping = {};
          _context.next = 3;
          return regeneratorRuntime.awrap((0, _allPages["default"])('Fetching authors', wp.users()));

        case 3:
          resources = _context.sent;
          tick = progress('Creating authors', resources.length);
          _iteratorNormalCompletion = true;
          _didIteratorError = false;
          _iteratorError = undefined;
          _context.prev = 8;
          _iterator = resources[Symbol.iterator]();

        case 10:
          if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
            _context.next = 18;
            break;
          }

          author = _step.value;
          _context.next = 14;
          return regeneratorRuntime.awrap(tick(author.name, dato.items.create({
            itemType: itemTypeId,
            name: author.name,
            slug: author.slug,
            username: author.username,
            firstName: author.first_name,
            lastName: author.last_name,
            email: author.email,
            url: author.url,
            description: author.description,
            nickname: author.nickname
          })));

        case 14:
          mapping[author.id] = _context.sent.id;

        case 15:
          _iteratorNormalCompletion = true;
          _context.next = 10;
          break;

        case 18:
          _context.next = 24;
          break;

        case 20:
          _context.prev = 20;
          _context.t0 = _context["catch"](8);
          _didIteratorError = true;
          _iteratorError = _context.t0;

        case 24:
          _context.prev = 24;
          _context.prev = 25;

          if (!_iteratorNormalCompletion && _iterator["return"] != null) {
            _iterator["return"]();
          }

        case 27:
          _context.prev = 27;

          if (!_didIteratorError) {
            _context.next = 30;
            break;
          }

          throw _iteratorError;

        case 30:
          return _context.finish(27);

        case 31:
          return _context.finish(24);

        case 32:
          return _context.abrupt("return", mapping);

        case 33:
        case "end":
          return _context.stop();
      }
    }
  }, null, null, [[8, 20, 24, 32], [25,, 27, 31]]);
}
